csrc/id_model/id_model.cpp

```diff
@@ -948,4 +956,265 @@ ValGraph IdModel::buildIntersection(
   return intersection;
 }
 
+namespace {
+
+// Check if there's an equivalent expression as iel_expr that uses
+// maybe_promoted_inputs. This is used to avoid redundantly replaying
+// expressions.
+// NOTE: This is currently overly conservative and some
+// opportunities for reuse are lost, althought it doesn't affect
+// the correctness of the analysis.
+Expr* findMatchingExpr(
+    const ExprGroup& iel_expr,
+    const ValGraph& iel_graph,
+    const std::vector<IterDomain*>& maybe_promoted_inputs) {
+  // Grab all uses of the promoted inputs
+  ExprGroups maybe_promoted_input_uses;
+  for (auto inp_id : maybe_promoted_inputs) {
+    // inp_id may have been just replayed, in which case it should
+    // not exist in the IEL graph. It should be just ignored as it
+    // should not have any use yet.
+    if (!iel_graph.hasGroup(inp_id)) {
+      continue;
+    }
+    const auto& inp_exact_group = iel_graph.toGroup(inp_id);
+    maybe_promoted_input_uses.pushBack(iel_graph.getUses(inp_exact_group));
+  }
+
+  // Look for exprs that have inputs that are mapped in the IEL
+  // graph with the (promoted) inputs of iel_expr.
+  for (const ExprGroup& maybe_promoted_input_use_group :
+       maybe_promoted_input_uses) {
+    NVF_ERROR(!maybe_promoted_input_use_group->empty());
+    // TODO: why skip this? If iel_expr is also an use of the promoted
+    // inputs, shouldn't it be also a candidate?
+    if (iel_expr == maybe_promoted_input_use_group) {
+      continue;
+    }
```

Comment:
I think this is correct:

> Should this be NVF_ERROR(iel_expr != maybe_promoted_input_use_group)?

Reply:
> Or... maybe... What happens to a never-concretized broadcasting ID? Is it promoted to another broadcasting ID, or it is never promoted?

If not concretized, it should not have any root resolution at Step 1, so there should be nothing to do.

Reply:
Actually, `iel_expr` can be equal to `maybe_promoted_input_use_group`.

For example:

```
e1: i0, b1 -> i0*b1
e2: i2, i3 ->i2*i3
```

And assume `b1` is promoted to `i3`. When `findMatchingExpr` is used for `e1` with promoted inputs of `i0` and `i3`, it would also find `e1` as a use of the promoted inputs in addition to `e2`, and obviously we would need to filter it out and pick `e2`, but we can't just do the assertion.

Reply:
I see. So conceptually
```C++
maybe_promoted_input_uses.pushBack(iel_graph.getUses(inp_iel_group));
```
is more natural to be
```C++
maybe_promoted_input_uses.computeIntersection(iel_graph.getUses(inp_iel_group));
```
but in practice, because we are checking same op and inputs mapped, it makes no difference.

Reply:
Yes, that's right.

Reply:
If that's the case, should we do
```C++
auto inp_id = maybe_promoted_inputs.at(0);
```
instead of
```C++
for (auto inp_id : maybe_promoted_inputs) {
```

Reply:
Not sure what you mean. Are you suggesting to look at the first ID only? That may not be actually a promoted input (that's why it's "maybe"), so that doesn't seem to work.

Reply:
Oh, I see what you mean. Yes, I think you're right. Let me change the code.

Reply:
https://github.com/NVIDIA/Fuser/pull/1777/commits/1a6c66d4a026b5a012ef0f94e9109e5b9ef886e7